lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,132 @@
+        if version_data:
+            last_supported_date = version_data[self.UPGRADE_DEADLINE_HEADER]
+            if last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+                if datetime.now().replace(tzinfo=UTC) > last_supported_date:
+                    return HttpResponse(status=426)
+
+    def process_response(self, _request, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < expiry of that version
+        """
+        version_data = self.get_info_for_version(_request)
+        if version_data:
+            upgrade_deadline = version_data[self.UPGRADE_DEADLINE_HEADER]
+            if upgrade_deadline != self.NO_LAST_SUPPORTED_DATE:
+                response[self.UPGRADE_DEADLINE_HEADER] = upgrade_deadline.isoformat()
+            latest_version = version_data[self.LATEST_VERSION_HEADER]
+            user_app_version = version_data[self.USER_APP_VERSION]
+            if (latest_version != self.NO_LATEST_VERSION and
+                    parsed_version(user_app_version) < parsed_version(latest_version)):
+                response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
+
+    def get_cache_key_name(self, user_agent, field):
+        """ get key name to use to cache any property against user agent """
+        return "mobile_api.app_version_upgrade.{}.{}".format(field, user_agent)
+
+    def get_info_for_version(self, request):
+        """
+        It sets request cache data for last_supported_date and latest_version with memcached values if exists against
+        user-agent else computes the values for specific platform and sets it in both memcache (for next
+        server interaction from same user-agent/platform) and request cache
+        """
+        user_agent = request.META.get('HTTP_USER_AGENT')
+        if user_agent:
+            request_cache_dict = RequestCache.get_request_cache().data
+            last_supported_date = request_cache_dict.get(self.UPGRADE_DEADLINE_HEADER, None)
+            latest_version = request_cache_dict.get(self.LATEST_VERSION_HEADER, None)
+            user_app_version = request_cache_dict.get(self.USER_APP_VERSION, None)
+            if last_supported_date and latest_version and user_app_version:
+                return {
+                    self.UPGRADE_DEADLINE_HEADER: last_supported_date,
+                    self.LATEST_VERSION_HEADER: latest_version,
+                    self.USER_APP_VERSION: user_app_version,
+                }
+
+            platform = self.get_platform(request, user_agent)
+            if platform:
+                request_cache_dict[self.USER_APP_VERSION] = platform.version
+                cached_data = cache.get_many([
+                    self.get_cache_key_name(user_agent, self.UPGRADE_DEADLINE_HEADER),
```

Comment:
Was there a real performance need to have an entry per user_agent in memcache?  We generally don't like to prematurely optimize for performance, unless we have data that shows it was needed.
Otherwise, premature optimizations add unnecessary complexity, can introduce stale data, and adds to our storage costs.